### PR TITLE
Updates to support docker development and production

### PR DESCRIPTION
**Pull request by jhnlsn**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -13,25 +13,25 @@ RUN npm install
 
 COPY runit /etc/service/node-app
 RUN chmod -R 755 /etc/service/node-app
-RUN npm install -g supervisor
+RUN npm install -g nodemon
 
 # Set correct environment variables.
```
Comment (inadarei) [inline Dockerfile]: Remove this commented-out leftover

File: Dockerfile [hunk 2]
```diff
@@ -13,25 +13,25 @@ RUN npm install
...
 # ENV HOME /opt/application
 
 EXPOSE 3000
 
-ENV NODE_PATH="/opt/application/lib" \
-    NODE_CONFIG_DISABLE_FILE_WATCH="Y" \
-    NODE_LOGGER_LEVEL="warning" \ 
+ENV NODE_APP_NAME=api \
```
Comment (inadarei) [inline Dockerfile]: Remove NODE_APP_NAME

File: Dockerfile [hunk 3]
```diff
@@ -13,25 +13,25 @@ RUN npm install
 
 COPY runit /etc/service/node-app
 RUN chmod -R 755 /etc/service/node-app
-RUN npm install -g supervisor
+RUN npm install -g nodemon
 
 # Set correct environment variables.
 # ENV HOME /opt/application
```
Comment (inadarei) [inline Dockerfile]: Add a comment explaining that the default values in Dockerfile are relevant to production. In dev we will always be using docker-compose, anyway so values there will override the ENV values in Dockerfile. By having Dockerfile values be sensible production values, you can easily run the container in prod, without using docker-compose (which is nice)

File: Dockerfile [hunk 4]
```diff
@@ -13,25 +13,25 @@ RUN npm install
...
 EXPOSE 3000
 
-ENV NODE_PATH="/opt/application/lib" \
-    NODE_CONFIG_DISABLE_FILE_WATCH="Y" \
-    NODE_LOGGER_LEVEL="warning" \ 
+ENV NODE_APP_NAME=api \
+    HOME_DIR=/opt/application \
+    NODE_PATH=/opt/application/lib \
+    NODE_CONFIG_DISABLE_FILE_WATCH=Y \
+    NODE_LOGGER_LEVEL=warning \
```
Comment (inadarei) [inline Dockerfile]: In production, this should probably be error?

File: Dockerfile [hunk 5]
```diff
@@ -13,25 +13,25 @@ RUN npm install
...
     NODE_LOGGER_GRANULARLEVELS=0 \
-    NODE_LOGGER_PLUGIN="util" \
-    NODE_LAUNCH_SCRIPT="/opt/application/server.js" \
+    NODE_LOGGER_PLUGIN=util \
     NODE_ENV=development \
```
Comment (inadarei) [inline Dockerfile]: Should be production

File: Dockerfile [hunk 6]
```diff
@@ -13,25 +13,25 @@ RUN npm install
...
     NODE_CLUSTERED=1 \
-    NODE_SERVE_STATIC=1 \
     NODE_HOT_RELOAD=1 \
```
Comment (inadarei) [inline Dockerfile]: Should be off, in production

File: docker-compose.yml [hunk 7]
```diff
@@ -1,6 +1,21 @@
-api: 
-  build: .
-  volumes:
-    - .:/opt/application  
-  ports:
-   - "5000:3000"
+version: '2'
+services:
+  api:
```
Comment (inadarei) [inline docker-compose.yml]: rename to "app"

File: docker-compose.yml [hunk 8]
```diff
@@ -1,6 +1,21 @@
...
+    build: .
+    volumes:
+      - .:/opt/application
+    ports:
+     - "5000:3000"
+    environment:
+      - HOME_DIR=/opt/application
+      - NODE_ENV=development
+      - NODE_PATH=/opt/application/lib
```
Comment (inadarei) [inline docker-compose.yml]: Try removing this both here and in Dockerfile and I think it will still work.

File: docker-compose.yml [hunk 9]
```diff
@@ -1,6 +1,21 @@
...
+      - NODE_CONFIG_DISABLE_FILE_WATCH=Y
+      - NODE_LOGGER_LEVEL=info
+      - NODE_LOGGER_GRANULARLEVELS=1
+      - NODE_LOGGER_PLUGIN=util
+      - NODE_CONFIG_DIR=/opt/application/config
+      - NODE_LOG_DIR=/opt/application/logs
+      - NODE_HOT_RELOAD=1
+      - NODE_CLUSTERED=0
```
Comment (inadarei) [inline docker-compose.yml]: Put it to "1" even in dev and let's see if that becomes pain in the neck.

File: package.json [hunk 10]
```diff
@@ -38,11 +38,19 @@
...
+    "docker-up": "docker-compose up -d",
+    "docker-clean-up": "npm run docker-stop && npm run docker-build && npm run docker-up",
+    "docker-stop": "docker-compose down && docker-compose rm -f",
+    "docker-restart": "docker-compose restart",
+    "logs": "docker-compose logs",
+    "test": "export NODE_PATH=${PWD}/lib; mocha --bail test/ test/acceptance/",
     "pretest": "jshint lib test",
-    "test-cov": "node_modules/istanbul/lib/cli.js cover node_modules/mocha/bin/_mocha -- --reporter dot test/ test/acceptance/",
-    "test-travis": "istanbul cover _mocha --report lcovonly -- test/ test/acceptance/"
+    "test-cov": "export NODE_PATH=${PWD}/lib; node_modules/istanbul/lib/cli.js cover node_modules/mocha/bin/_mocha -- --reporter dot test/ test/acceptance/",
```
Comment (inadarei) [inline package.json]: 1. I think you can remove "export NODE_PATH". Please test
2.Add docker-test and docker-test-cov so you can run tests inside the container. Right now you can only run them on the host

File: runit/run [hunk 11]
```diff
@@ -1,6 +1,5 @@
 #!/bin/sh
 # see: https://github.com/phusion/baseimage-docker#adding-additional-daemons
```
Comment (inadarei) [inline runit/run]: Remove this. It serves no purpose, anymore

**Comment by inadarei** [comment]
Please also refresh dev_start.sh and start.sh so
1. They use nodemon
2. There's some unnecessary code in there that can be removed.

**Comment by jhnlsn** [comment]
TODO

1. update start.sh and dev_start.sh
1. run tests inside the docker container

**Comment by jhnlsn** [comment]
@inadarei can we merge in this pull request to support docker and then have a look at the local start/dev_start in another PR?

**Pull request merged**